Refactor autoscaler config

AdamGleave opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?

The current config format consists of `initialization_commands`, `setup_commands` and `*_start_ray_commands`. The format is inconsistent, with some but not all having variants for head and worker nodes, and the distinction between `initialization_commands` and `setup_commands` is blurry.

This PR refactors the config to consist of three entries:
  - `setup_commands`, run when the node is created.
  - `boot_commands`, whenever the node is powered on.
  - `start_ray_commands`, whenever Ray is (re)started.

Each of these entries is a dict that can contain keys `common`, `head` and `worker`. `common` is always executed first, followed by the node-specific commands, if specified.

A particular pain point, described in #6128, is that node re-use is incompatible with Docker. This PR fixes this by giving the user control over which commands get executed in Docker at different stages, by use of an `IN_DOCKER` macro.

## Related issue number

Fixes #6128. The changes broadly follow the format agreed in the discussion; the main difference is I opted for a hierarchical config format rather than creating `{head,worker}_setup_commands`, `{head,worker}_restart_commands`, etc.

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

AmplabJenkins commented:
Can one of the admins verify this patch?

AdamGleave commented:
@ericl @richardliaw @hartikainen I'm not sure how to determine whether a node has just been created, or whether a stopped cached node is being reused. Any ideas? I need this to determine whether or not to run `setup_commands`.

Besides this (which is causing one test to fail) I believe the PR is ready.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18555/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18556/
Test PASSed.

hartikainen commented on python/ray/autoscaler/autoscaler.py:
```diff
@@ -130,32 +141,27 @@
-    # Commands that will be run on the head node after common setup.
-    "head_setup_commands": (list, OPTIONAL),
+    # Commands that will be run when the cluster is first created.
+    "setup_commands": COMMANDS_SCHEMA,
```
Should we call these:
- `node_setup_commands`
- `node_restart_commands`
- `ray_restart_commands`
?

AdamGleave replied:
Good suggestion, changed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18585/
Test FAILed.

edoakes commented on python/ray/autoscaler/docker.py:
```diff
@@ -50,31 +65,38 @@ def dockerize_if_needed(config):
+    res = []
+    for cmd in cmds:
+        args = cmd.split(" ")
+        if not args[0] == "IN_DOCKER":
```
this is really ugly and I don't like the idea of having magic keywords like this in the commands in general. what is the use case for this? why wouldn't these commands either be run in the dockerfile or as part of the application?

AdamGleave replied:
When using Docker with Ray, we need to run most commands inside the Docker container; e.g. `ray start`.

Currently this wrapping happens implicitly for (what was) `*_setup_commands` and `*_start_ray_commands`. This is problematic: some commands need to run outside of Docker (like mounting EFS instances on AWS), some commands need to run inside Docker (like updating source code of your application). Previously one could use `*_initialization_commands` to run things outside Docker. However, with the recently added support for caching stopped nodes this broke as `initialization_commands` will not get run on node restart.

See https://github.com/ray-project/ray/issues/6128 for discussion that led to this API.

edoakes replied:
The `ray start` command should always be run in the container when using docker so we don't need this syntax for that case.

Seems like if you're using docker the "right" way to do this is to push a new image, but I can see how that would be annoying/high overhead.

Would this be fixed by running always running `node_start_commands` inside of docker? Or do you think there are cases where people need to run commands on both the host and in the container on restart?

AdamGleave replied:
I think sometimes people need to run commands on the host on restart. For example, mounting external file systems is awkward to do inside of Docker (EC2 credentials may not automatically be passed through).

Good point about the canonical Docker approach being to update source code by pushing a new image. I think this would also require change in Ray's design, though, since right now the Docker container persists until you reboot the node.

I'd be OK removing the `ray start` commands and having a separate config, but think there should be some functionality of running commands each time `ray` is restarted.

AdamGleave commented on python/ray/autoscaler/autoscaler.py:
```diff
@@ -130,32 +141,27 @@
-    # Command to start ray on worker nodes. You shouldn't need to modify this.
-    "worker_start_ray_commands": (list, OPTIONAL),
+    # Commands that will be run whenever Ray is (re)started.
+    "ray_restart_commands": COMMANDS_SCHEMA,
```
There seems to be a lot of heterogenity in the example Ray configs for how Ray is started. For example, `aws/example-full.yaml` sets a limit on the number of open files; `local/example-full.yaml` disables core dump limit (presumably for debugging). In my own usage I've often used the `--num-cpus` flag to reduce the number of reported CPUs on the head node to ensure it isn't overloaded.

Maybe `ray_head_config` and `ray_worker_config` could capture this, but I think it'd be better for you to implement that as I don't have a clear vision for what this would look like.

AdamGleave replied:
I also have sometimes put commands like `git pull` in this section so that I can just run `ray up` to restart the cluster with a new version of the code.

edoakes replied:
Number of open files & core dump can and should should be specified elsewhere and overloading `ray up` with `git pull` commands isn't the right API. I would strongly prefer removing these and replacing them with config fields as part of this PR or before merging this because I don't want to make two breaking API changes..

AdamGleave replied:
Agree makes sense to consolidate all breaking API changes into one release. I'm fine with eliminating the `ray start` command and replacing it with a config, but I'm not that familiar with how `ray start` works so would appreciate some pointers here. Some options I can see:
  1. Auto-construct the `ray start` command-line, and have the user optionally specify additional CLI flags to append (defaults to `""`).
  2. Have the user specify a YAML dict of flags to `ray start`. Some if unspecified are automatically filled in with appropriate values (e.g. `redis-port`). Then construct CLI flag from that.
  3. More restricted version where we decide what CLI flags to expose in the autoscaler.

ericl replied:
Why change the ray start commands? I think having it explicitly here is the simplest solution. Understanding configurations is not easy for users to do, and you may run into situations that are not supported by the config.

AdamGleave commented on python/ray/autoscaler/autoscaler.py:
```diff
@@ -39,6 +39,17 @@
+COMMANDS_SCHEMA = (
+    {
+        # Setup common to all nodes. These are run first.
+        "common": (list, OPTIONAL),
```
Feels a bit verbose to me given they're subfields, so it'd be e.g.:
```
node_restart_commands:
  common_setup_commands: []
```

I'd be OK with `*_commands`.

ericl replied:
I think we should avoid nesting here (which can be confusing for new users), and have all the combinations materialized explicitly. What about?
```
setup_commands <--- majority of users will use just this
head_setup_commands
worker_setup_commands

start_services_commands (suggesting "start_services" this as a replacement for 'restart')
head_start_services_commands
worker_start_services_commands

head_start_ray_commands
worker_start_ray_commands
```
Note that for the most part, this preserves compatibility with existing configuration files: setup commands remains the same, and we are just removing init commands.

AdamGleave replied:
Done (except I used `node_start_commands` as suggested by edoakes in other thread).

AdamGleave commented on python/ray/autoscaler/autoscaler.py:
```diff
@@ -130,32 +141,27 @@
-    # Commands that will be run on worker nodes after common setup.
-    "worker_setup_commands": (list, OPTIONAL),
+    # Commands that will be run whenever a node is booted.
+    "node_restart_commands": COMMANDS_SCHEMA,
```
Yes, I've amended the comment accordingly. I'm OK renaming it but agree `node_start_or_restart_commands` is a bit verbose. Some alternatives:
  - `node_boot_commands`
  - `node_start_commands`
  - `node_poweron_commands`

edoakes replied:
`node_start_commands` sounds good

ericl replied:
So to summarize, we have `node_setup` -> `node_start` -> `ray_start`?

AdamGleave replied:
Done

AdamGleave commented on python/ray/autoscaler/autoscaler.py:
```diff
@@ -130,32 +141,27 @@
 }
 
 
+def get_commands(config, key, head):
```
good point, renamed

AdamGleave commented on python/ray/autoscaler/autoscaler.py:
```diff
@@ -559,13 +565,16 @@ def _update(self):
-                                             for node_id in nodes):
+        for node_id in nodes:
+            item = self.should_update(node_id)
+            (node_id, node_setup_commands, node_restart_commands,
```
good point, think I did this to avoid wrapping, but then couldn't get it to fit. have reformatted.

AdamGleave commented on python/ray/autoscaler/autoscaler.py:
```diff
@@ -130,32 +141,27 @@
 
 
+def get_commands(config, key, head):
+    """Get commands for a head or worker node under key `key`."""
```
expanded doc

AdamGleave commented on python/ray/autoscaler/autoscaler.py:
```diff
@@ -663,10 +674,10 @@ def recover_if_needed(self, node_id, now):
+            node_setup_commands=[],
+            node_restart_commands=[],
+            ray_restart_commands=with_head_node_ip(
+                get_commands(self.config, "ray_restart_commands", head=False)),
```
I'm unsure about this. It would make renaming things easier (can use IDEs to automatically find usage), but I assume we don't want to do that often as it will break users configs? There are also a bunch of uses it won't capture (e.g. in example YAML files), so a find-and-replace would still be necessary.

If you still feel this is worth doing I'm OK making the change, though.

edoakes replied:
Better to have it as consolidated as possible IMO. At least this would mean only looking for it/changing it one place in the code itself (and next to the schema instead of scattered throughout the codebase), changing YAML files is unavoidable without some sort of templating magic.

AdamGleave replied:
Done

AdamGleave commented:
Thanks for the review @edoakes. I've made the documentation and variable naming changes suggested. There are a few other areas where some discussion is still needed.

Also, in case it got lost, do you have any suggestions for how to determine whether a node has just been created, or whether it is a stopped cached node is being reused? I can't see a way to do this with the current information we're storing, but this is needed determine whether or not to run setup_commands.

ericl commented:
> how to determine whether a node has just been created, or whether it is a
stopped cached node is being reused

You could potentially do this just in the config, by touching a file at end
of setup (e.g., if [ -e /tmp/SETUP_OK ]; then stuff; fi)

Eric


On Wed, Nov 20, 2019 at 3:40 PM Adam Gleave <notifications@github.com>
wrote:

> Thanks for the review @edoakes <https://github.com/edoakes>. I've made
> the documentation and variable naming changes suggested. There are a few
> other areas where some discussion is still needed.
>
> Also, in case it got lost, do you have any suggestions for how to
> determine whether a node has just been created, or whether it is a stopped
> cached node is being reused? I can't see a way to do this with the current
> information we're storing, but this is needed determine whether or not to
> run setup_commands.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/6182?email_source=notifications&email_token=AAADUSQZXBBMVXF4W7ZBTJDQUXDG5A5CNFSM4JOM2HL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEWEW3Y#issuecomment-556551023>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAADUSSZNVA6AXGSLDXSQTLQUXDG5ANCNFSM4JOM2HLQ>
> .
>

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18650/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18655/
Test PASSed.

edoakes commented:
Left comments on what I think the outstanding issues are. Regarding detecting if a node is being restarted, I would suggest storing a tag that indicates if a node was restarted rather than created and then a new call to `NodeProvider` (that's only implemented for AWS) that returns that information. The call should only be made if `cache_stopped_nodes=True`, so that only `NodeProvider`s that implement that flag need to implement it.

ericl commented:
What are the new proposed semantics? Can we summarize those first before proceeding?

AdamGleave commented:
I've added support to not rerun `node_initalization_commands` for cached nodes but to still run `node_setup_commands`. I've also made the requested renames of the config variables.

The outstanding issues I am aware of:
  - Whether to use `IN_DOCKER` macro and, if not, what alternative to use in the config;
  - Whether to construct `ray start` commands from a structured config, rather than free-form `ray_restart_commands`.

ericl commented:
Ok, looking at the new changes and the comments, I think we can make a few tweaks:

```

# List of commands that will be run when the node is first created.
setup_commands:
head_setup_commands:
worker_setup_commands:

# List of commands that will be run each time the node is booted.
# On the first boot, these will run after `setup_commands`.
node_start_commands: []
head_node_start_commands: []
worker_node_start_commands: []

# Commands to start Ray on the nodes. You do not need to change this.
head_start_ray_commands:
worker_start_ray_commands:
```

The above config (1) preserves full backwards compatibility for non-docker users, and (2) should support the full feature set required. Any reason not to adopt this exact spec @edoakes ?

AdamGleave commented:
> Ok, looking at the new changes and the comments, I think we can make a few tweaks:
> 
> ```
> 
> # List of commands that will be run when the node is first created.
> setup_commands:
> head_setup_commands:
> worker_setup_commands:
> 
> # List of commands that will be run each time the node is booted.
> # On the first boot, these will run after `setup_commands`.
> node_start_commands: []
> head_node_start_commands: []
> worker_node_start_commands: []
> 
> # Commands to start Ray on the nodes. You do not need to change this.
> head_start_ray_commands:
> worker_start_ray_commands:
> ```
> 
> The above config (1) preserves full backwards compatibility for non-docker users, and (2) should support the full feature set required. Any reason not to adopt this exact spec @edoakes ?

So the difference is renaming `node_setup_commands` to `setup_commands` and `ray_restart_commands` to `ray_start_commands`?

I think @hartikainen wanted `node_setup_commands` for consistency with `node_start_commands`.

edoakes commented:
@ericl and I discussed more offline and this seems good, also ok with the `IN_DOCKER` macro. A few naming changes suggested below for clarity/consistency:

```
# List of commands that will be run when the node is first created.
# Head/worker node commands are run after common commands.
node_creation_commands:
head_node_creation_commands: []
worker_node_creation_commands: []

# List of commands that will be run each time the node is booted.
# Head/worker node commands are run after common commands.
# On the first boot, these will run after `creation_commands`.
node_start_commands: []
head_node_start_commands: []
worker_node_start_commands: []

# Command that is run to start Ray on the nodes. You should not need to change this.
head_start_ray_command:
worker_start_ray_command:
```

ericl commented:
That looks good. I would also retain backwards compatibility aliases for
the old setup commands, but update the docs to use the new node creation
commands name.

On Fri, Nov 22, 2019, 3:49 PM Edward Oakes <notifications@github.com> wrote:

> @ericl <https://github.com/ericl> and I discussed more offline and this
> seems good, also ok with the IN_DOCKER macro. A few naming changes
> suggested below for clarity/consistency:
>
> # List of commands that will be run when the node is first created.
> # Head/worker node commands are run after common commands.
> node_creation_commands:
> head_node_creation_commands: []
> worker_node_creation_commands: []
>
> # List of commands that will be run each time the node is booted.
> # Head/worker node commands are run after common commands.
> # On the first boot, these will run after `creation_commands`.
> node_start_commands: []
> head_node_start_commands: []
> worker_node_start_commands: []
>
> # Command that is run to start Ray on the nodes. You should not need to change this.
> head_start_ray_command:
> worker_start_ray_command:
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/6182?email_source=notifications&email_token=AAADUSWM6EMCO7BF5TSIYXLQVBVYJA5CNFSM4JOM2HL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEE7GBIQ#issuecomment-557736098>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAADUSWK4ZCEJT3HKLWOBU3QVBVYJANCNFSM4JOM2HLQ>
> .
>

AdamGleave commented:
OK, I've made the requested naming changes and added backward compatibility logic.

edoakes commented:
Excellent. Could you also make sure all of the docs are up to date? You can find the source under `doc/source` and the primary page you care about is `autoscaling.rst` (but also check out the k8s docs please). You can build them locally by doing `pip install doc/requirements-doc.txt` and then `make html` from within the `doc` directory.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18772/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18775/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18782/
Test PASSed.

AdamGleave commented:
> Excellent. Could you also make sure all of the docs are up to date? You can find the source under `doc/source` and the primary page you care about is `autoscaling.rst` (but also check out the k8s docs please). You can build them locally by doing `pip install doc/requirements-doc.txt` and then `make html` from within the `doc` directory.

I've checked and can't see anywhere the docs need changing. They don't go into much detail on the autoscaler config, instead referring to example YAML files. Given this I've updated the comments in the YAML files to be more explanatory and consistent between the different setups (AWS, GCP, local, etc).

I also fixed a typo which was causing some tests to fail. All `test_autoscaler.py` and `test_autoscaler_yaml.py` tests now passing on my machine

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18808/
Test FAILed.

AdamGleave commented:
The failing tests seem to be unrelated to my changes. In Travis 2.7 passes on Linux, 3.5 hangs on `test_dependencies`. Jenkins times out on a Ray tune test that these changes should not have effected.

ericl commented on python/ray/autoscaler/aws/example-gpu-docker.yaml:
```diff
@@ -103,25 +103,22 @@ file_mounts: {
+node_start_commands:
     # - pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.8.0.dev6-cp27-cp27mu-manylinux1_x86_64.whl
-    - pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.8.0.dev6-cp35-cp35m-manylinux1_x86_64.whl
+    - IN_DOCKER pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.8.0.dev6-cp35-cp35m-manylinux1_x86_64.whl
```
To make it more clear IN_DOCKER is a magic command, how about $IN_DOCKER instead (or something else if it doesn't really have env var semantics).

AdamGleave replied:
It doesn't quite have environment variable semantics since the following string gets *quoted*, and then the `docker exec ...` command is prepended. In this sense it feels more like a command or a bash function to me, so I'm uneasy about `$IN_DOCKER` (also would we then need to handle alternative expansions, like `${IN_DOCKER}`?)

Certainly open to alternatives though. Perhaps adding a special symbol would help, e.g. `@IN_DOCKER`? Or even just a more explicit name like `RUN_IN_DOCKER` or `DOCKER_EXEC`.

richardliaw replied:
@IN_DOCKER sounds good.

ericl commented on python/ray/autoscaler/aws/node_provider.py:
```diff
@@ -222,7 +226,9 @@ def create_node(self, node_config, tags, count):
                     InstanceIds=reuse_node_ids)
+                reuse_tags = dict(tags)
                 for node_id in reuse_node_ids:
+                    reuse_tags[TAG_RAY_FIRST_BOOT] = False
```
The reuse tags dict doesn't seem to be ever used.

AdamGleave replied:
thanks, this is a bug

ericl commented on python/ray/autoscaler/updater.py:
```diff
@@ -414,17 +416,17 @@ def do_update(self):
 
-            with LogTimer(self.log_prefix + "Setup commands completed"):
-                for cmd in self.setup_commands:
+        if not same_config or self.provider.is_cached(self.node_id):
```
Is this correct? Doesn't it mean we will always run these commands every time the node is updated, if it is a cached node?

AdamGleave replied:
It's intended that `start_commands` should run every time the node is booted (even if it is cached). Is the concern that running e.g. `ray up` will cause this to be rerun? That might be an issue, I'm not sure how to distinguish those two cases.

ericl replied:
Yeah, since it's rerun on ray up its behaviour is almost identical to the Ray start commands.

Would it actually be reasonable to drop this and use ray start commands for node start commands instead?

AdamGleave commented:
Sometimes you do want to run a command once per boot (e.g. mounting EFS
file systems). Can always get around it by writing idempotent commands, but
may make it harder for the user.

On Sat, 23 Nov 2019 at 20:57, Eric Liang <notifications@github.com> wrote:

> *@ericl* commented on this pull request.
> ------------------------------
>
> In python/ray/autoscaler/updater.py
> <https://github.com/ray-project/ray/pull/6182#discussion_r349903122>:
>
> >                      self.cmd_runner.run(cmd)
>
> -            with LogTimer(self.log_prefix + "Setup commands completed"):
> -                for cmd in self.setup_commands:
> +        if not same_config or self.provider.is_cached(self.node_id):
>
> Yeah, since it's rerun on ray up its behaviour is almost identical to the
> Ray start commands.
>
> Would it actually be reasonable to drop this and use ray start commands
> for node start commands instead?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/6182?email_source=notifications&email_token=AALZ3IZTPRT6GVJIIATU26TQVICSBA5CNFSM4JOM2HL2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCMYHAFI#discussion_r349903122>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALZ3I7OSCZEE37FX7WSCWDQVICSBANCNFSM4JOM2HLQ>
> .
>

ericl commented:
> Sometimes you do want to run a command once per boot (e.g. mounting EFS
file systems). Can always get around it by writing idempotent commands, but
may make it harder for the user.

Agreed it is harder. However, my concern is that the current implementation doesn't actually work for cached nodes (which is the common case for cloud users), so it degrades to once per update anyways.

The options I see are (1) fixing the node state handling so it is once per boot in both cases, or (2) replacing this feature with an example like this:

`[ -e /tmp/boot_ok ] || <your first boot command> && touch /tmp/boot_ok`

Maybe this can be made available in a similar way as IN_DOCKER, i.e.:
`ON_NODE_START_ONLY <your first boot command>`.

I strongly prefer (2) since it reduces the complexity of the ray implementation.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18820/
Test PASSed.

AdamGleave commented:
> > Sometimes you do want to run a command once per boot (e.g. mounting EFS
> > file systems). Can always get around it by writing idempotent commands, but
> > may make it harder for the user.
> 
> Agreed it is harder. However, my concern is that the current implementation doesn't actually work for cached nodes (which is the common case for cloud users), so it degrades to once per update anyways.
> 
> The options I see are (1) fixing the node state handling so it is once per boot in both cases, or (2) replacing this feature with an example like this:
> 
> `[ -e /tmp/boot_ok ] || <your first boot command> && touch /tmp/boot_ok`
> 
> Maybe this can be made available in a similar way as IN_DOCKER, i.e.:
> `ON_NODE_START_ONLY <your first boot command>`.
> 
> I strongly prefer (2) since it reduces the complexity of the ray implementation.

(2) does seem the simplest solution. I have a general hesitation about pushing too much complexity into configs and especially via bash scripts which are generally harder to maintain than Python. But in this case I don't see much downside.

This would mean more of a breaking API change: we'd be deleting `node_setup_commands` entirely. Migth also want to rename `start_ray_commands`, since it'd be doing double-duty, e.g. to just `start_commands`.

To get a sense of what (1) would be like, the most direct approach I can see would be to add another tag to nodes indicating if `node_start_commands` has been run. There may be a cleaner solution though: I still don't have a great mental model of the `autoscaler`.

Perhaps @edoakes or @hartikainen have thoughts on this? Once there's consensus I'm happy to make either change.

ericl commented:
Ah, I was thinking you would still keep the node setup commands, but only
remove the start ones (which were added in this PR anyways).

So for starting EFS, that would be a little more complicated but you could
put it in the (Ray) start commands. Any remaining work like installing
dependencies could still be left in setup, where it would only be rerun if
the config is updated.

On Sun, Nov 24, 2019, 10:43 AM Adam Gleave <notifications@github.com> wrote:

> *@AdamGleave* commented on this pull request.
> ------------------------------
>
> In python/ray/autoscaler/aws/example-gpu-docker.yaml
> <https://github.com/ray-project/ray/pull/6182#discussion_r349939649>:
>
> >      # - pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.8.0.dev6-cp27-cp27mu-manylinux1_x86_64.whl
> -    - pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.8.0.dev6-cp35-cp35m-manylinux1_x86_64.whl
> +    - IN_DOCKER pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.8.0.dev6-cp35-cp35m-manylinux1_x86_64.whl
>
> It doesn't quite have environment variable semantics since the following
> string gets *quoted*, and then the docker exec ... command is prepended.
> In this sense it feels more like a command or a bash function to me, so I'm
> uneasy about $IN_DOCKER (also would we then need to handle alternative
> expansions, like ${IN_DOCKER}?)
>
> Certainly open to alternatives though. Perhaps adding a special symbol
> would help, e.g. @IN_DOCKER? Or even just a more explicit name like
> RUN_IN_DOCKER or DOCKER_EXEC.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/6182?email_source=notifications&email_token=AAADUSQRX7EX6EQGU3Q7JHTQVLDOLA5CNFSM4JOM2HL2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCMYP43A#discussion_r349939649>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAADUSWH4EWMLDZMF6JKMWDQVLDOLANCNFSM4JOM2HLQ>
> .
>

richardliaw commented:
@edoakes maybe you can just confirm so we can merge this?

edoakes commented:
I'm good with Eric's suggestion: solution (2) above with leaving `node_setup_commands` and removing the start commands introduced in this PR to minimize the API change. It's a little ugly to do things in the ray start commands, but I like the simplicity of it.

AdamGleave commented:
Thanks, I'm glad we could reach consensus on this.

As requested, I have:
  - Removed `node_start_commands`
  - Added an `ON_NODE_START_ONLY` macro to provide similar functionality

One issue I ran into is that `docker run` needs to (a) run before any other `*start_ray_commands`, and (b) is different for head and worker node. We have so far been running `start_ray_commands` *before* `{head,worker}_start_ray_commands`, which is incompatible with. 

To fix this, I've flipped the order: we now execute `{head,worker}_start_ray_commands` before the common commands. This seems OK with all the example configs but may have some unintended consequences.

Do we want to rename `node_creation_commands` *back* to `setup_commands` now? I think the new name is clearer (and we do have some backward compatibility code to support `setup_commands` still), but there seems less need to rename it now there isn't `node_start_commands`.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19303/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19304/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19305/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19314/
Test FAILed.

edoakes commented:
`node_creation_commands` seems fine if we support the old name.

In what way is `docker run` different for the head and worker nodes? And how is running the common commands first incompatible? It's pretty counterintuitive to run the type-specific commands first and then the common commands - let's avoid that if at all possible.

AdamGleave commented:
> In what way is `docker run` different for the head and worker nodes?

The config supports different run options and image for the head and worker. This was added in https://github.com/ray-project/ray/pull/4537 @richardliaw @hartikainen participated in that PR so can perhaps comment on why these options exist and how important it is to retain them?

> And how is running the common commands first incompatible? It's pretty counterintuitive to run the type-specific commands first and then the common commands - let's avoid that if at all possible.

`docker.py` works by rewriting the config; it doesn't directly run any commands. So we have to add the `docker run` command to one of the existing config entries. Now that we've eliminated `node_start_commands`, the only places are `ray_start_commands` and `{head,worker}_ray_start_commands`.

`docker run` has to be run *before* any of the other Ray start commands. So we have to add it to whichever of the `*ray_start_commands` gets executed first.

One alternative would be to have `docker.py` delete the `ray_start_commands` entry entirely, and just splice it into an appropriate place in `{head,worker}_ray_start_commands`. Bit hacky though.

hartikainen commented:
> > In what way is `docker run` different for the head and worker nodes?
> 
> The config supports different run options and image for the head and worker. This was added in #4537 @richardliaw @hartikainen participated in that PR so can perhaps comment on why these options exist and how important it is to retain them?

One example is when the user wants to run GPU-less head node and GPU workers. In this case, both the image and run command could be different (i.e. the GPU command need something like `--runtime=nvidia` and the GPU images might need cuda/nvidia drivers etc installed).

AdamGleave commented:
> One example is when the user wants to run GPU-less head node and GPU workers. In this case, both the image and run command could be different (i.e. the GPU command need something like `--runtime=nvidia` and the GPU images might need cuda/nvidia drivers etc installed).

That seems important to keep.

richardliaw commented:
Hi @AdamGleave; sorry for the slow replies. Can you provide an example of what you're trying to do that is forcing you to reverse the common commands and the head/worker commands?

AdamGleave commented:
> Hi @AdamGleave; sorry for the slow replies. Can you provide an example of what you're trying to do that is forcing you to reverse the common commands and the head/worker commands?

This isn't config-specific: it always needs to happen when using Docker with the current setup.

`docker.py` works by rewriting the config. We need to insert a `docker run` *somewhere* in the config. It can't be in the `node_creation_commands` since it needs to be run on each boot. It can't be in `start_ray_commands` since the `docker run` can be different between head and worker. So this leaves `{head,worker}_start_ray_commands`.

Since commands in `start_ray_commands` may need to run inside Docker, this forces us to have `{head,worker}_start_ray_commands` run first.

richardliaw commented:
> Since commands in start_ray_commands may need to run inside Docker, this forces us to have {head,worker}_start_ray_commands run first.

Will all of these issues be resolved if we made it such that `docker run` must be explicitly done by the user? Then, for those using `docker`, they must be responsible for defining the order of commands.

AdamGleave commented:
> Will all of these issues be resolved if we made it such that `docker run` must be explicitly done by the user? Then, for those using `docker`, they must be responsible for defining the order of commands.

I think that would work. Constructing the `docker run` command may be slightly non-trivial (e.g. need to expose the right ports), but we could potentially provide a macro to help with this.

richardliaw commented:
Alright. Shall we move forward with this then? (i.e., making the user manually do the docker run, and also maintaining the consistent ordering of command execution). Happy to review when done.

AdamGleave commented:
@hartikainen @edoakes what're your thoughts on this? Happy to proceed once there's consensus on a design.

ericl commented:
> Do we want to rename node_creation_commands back to setup_commands now? I think the new name is clearer (and we do have some backward compatibility code to support setup_commands still), but there seems less need to rename it now there isn't node_start_commands.

Just a nit here, I do think it was clearer before, but now that we have ON_NODE_START_ONLY the original name seems more appropriate since it's not run on node creation but each repeated setup.

AdamGleave commented:
> > Do we want to rename node_creation_commands back to setup_commands now? I think the new name is clearer (and we do have some backward compatibility code to support setup_commands still), but there seems less need to rename it now there isn't node_start_commands.
> 
> Just a nit here, I do think it was clearer before, but now that we have ON_NODE_START_ONLY the original name seems more appropriate since it's not run on node creation but each repeated setup.

OK, I can revert that change.

Things like this make me really want to get consensus before writing any more code though... we've gone through a lot of iterations on the design in this PR already.

edoakes commented:
Forcing the user to explicitly run docker commands sounds good to me assuming that it wouldn't be too burdensome. @AdamGleave to avoid having to backtrack on anything again maybe you could provide a minimal sample config that uses manual `docker run` commands without actually implementing it first?

AdamGleave commented:
On reflection I think whether to keep `docker.py` or specify it in the config is an orthogonal decision to whether to run `{head,worker}_start_ray_commands` before `start_ray_commands`. In particular even a custom config will not be able to handle different Docker arguments for head/worker if `start_ray_commands` runs first: the only recourse currently would be for the user to just not specify `start_ray_commands` at all, and copy & paste it to the end of `{head,worker}_start_ray_commands`.

One alternative that would solve this is to let the user explicitly specify priorities, e.g:

```
start_ray_commands:
    - common: echo "this runs before ANYTHING"
      head: echo "this runs second"
    - common: echo "this runs after everything else"
```

Here's an example of a custom config, assuming the change introduced in this branch running `{head,worker}_` first is retained:

```
cluster_name: ray-on-docker
# note docker: section would be removed
provider:
    type: aws
    region: us-west-2
    availability_zone: us-west-2a
auth:
    ssh_user: ubuntu

head_start_ray_commands:
    - docker inspect -f '{{.State.Running}}' ray-container || docker run --rm --name ray-container -d -it -p 6379:6379 -p 8076:8076 -p 4321:4321 -v /host/path:/guest/path -e LC_ALL=C.UTF-8 -e LANG=C.UTF-8 --net=host --runtime=nvidia tensorflow/tensorflow:1.12.0-gpu-py3
    - IN_DOCKER ulimit -n 65536; ray start --head --redis-port=6379 --object-manager-port=8076 --autoscaling-config=~/ray_bootstrap_config.yaml
worker_start_ray_commands:
    - docker inspect -f '{{.State.Running}}' ray-container || docker run --rm --name ray-container -d -it -p 6379:6379 -p 8076:8076 -p 4321:4321 -v /host/path:/guest/path -e LC_ALL=C.UTF-8 -e LANG=C.UTF-8 --net=host tensorflow/tensorflow:1.12.0-py3
    - IN_DOCKER ulimit -n 65536; ray start --redis-address=$RAY_HEAD_IP:6379 --object-manager-port=8076
```

It's quite verbose. I could introduce another macro to try and cut it down but honestly it seems hard to make it reliable and flexible: Docker is sensitive to the order some arguments are placed for example.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19816/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19819/
Test PASSed.

richardliaw commented:
OK, hm to revisit your original issue here:

> One issue I ran into is that `docker run` needs to (a) run before any other *start_ray_commands, and (b) is different for head and worker node. We have so far been running start_ray_commands before {head,worker}_start_ray_commands, which is incompatible with.

As you've shown, the docker run command is too verbose. Sorry if this is redundant and please correct any misunderstandings you see - I'm under the impression that this PR should be doing the following. The docker run is placed after setup_commands because you may want to execute things on the raw AMI before docker is started. Note that in this case there is no "IN_DOCKER" allowed in setup_commands (and `setup_commands` essentially becomes the former `initialization_commands`). It is only in `start_ray_commands` where you can use the IN_DOCKER primitive in the `start_ray*`. 

(Note that common commands are not bulk executed across both workers and head at the same time, and this should also be the way how setup_commands is handled iirc):

```
[setup_commands on head] ->
[implicit docker run]@head -> 
start_ray_commands@head -> 
head_start_ray_commands@head -> 
[autoscale workers] -> 
[setup_commands on workers] -> 
[implicit docker run]@workers ->
start_ray_commands@worker -> 
worker_start_ray_commands@worker
```

Again, sorry if this is redundant! I'm just trying to regain context and get this merged...

AdamGleave commented:
> OK, hm to revisit your original issue here:
> 
> > One issue I ran into is that `docker run` needs to (a) run before any other *start_ray_commands, and (b) is different for head and worker node. We have so far been running start_ray_commands before {head,worker}_start_ray_commands, which is incompatible with.
> 
> As you've shown, the docker run command is too verbose. Sorry if this is redundant and please correct any misunderstandings you see - I'm under the impression that this PR should be doing the following. The docker run is placed after setup_commands because you may want to execute things on the raw AMI before docker is started. Note that in this case there is no "IN_DOCKER" allowed in setup_commands (and `setup_commands` essentially becomes the former `initialization_commands`). It is only in `start_ray_commands` where you can use the IN_DOCKER primitive in the `start_ray*`.
> 
> (Note that common commands are not bulk executed across both workers and head at the same time, and this should also be the way how setup_commands is handled iirc):
> 
> ```
> [setup_commands on head] ->
> [implicit docker run]@head -> 
> start_ray_commands@head -> 
> head_start_ray_commands@head -> 
> [autoscale workers] -> 
> [setup_commands on workers] -> 
> [implicit docker run]@workers ->
> start_ray_commands@worker -> 
> worker_start_ray_commands@worker
> ```
> 
> Again, sorry if this is redundant! I'm just trying to regain context and get this merged...

Sorry for the slow reply, had limited internet access past few weeks.

This is a good summary. Just to add that the `docker run` command often needs to be different between head/workers, so cannot be inserted into `start_ray_commands`.

The simplest approach to accommodate this if we don't want to edit the config format is to make `docker.py` rewrite the config to be `{head,worker}_start_ray_commands = [implicit docker run] + start_ray_commands + {head,worker}_start_ray_commands` and just delete `start_ray_commands` from the config. This may be a bit confusing for the user if they ever look at the config being sent across the wire, though.

richardliaw commented:
Yes! That sounds fine. I think we already do a bit of mutilation to the config, so it's probably OK to do as suggested.

AdamGleave commented:
> Yes! That sounds fine. I think we already do a bit of mutilation to the config, so it's probably OK to do as suggested.

@edoakes @hartikainen @ericl does this sound OK to you? If so I should be able to carve out time sometime next week (though may be busy with ICML deadline).

hartikainen commented:
Sounds good to me!

AdamGleave commented:
I'm busy with ICML right now, but will try and work on this w/c 10th Feb. (If anyone feels blocked on this then feel free to take over the PR to get it over the finish.)

edoakes commented:
This is pretty out of date, so going to close for now but please reopen this or a new pr if need be. Note that @ijrsvt has made a lot of progress on improving docker support.
